### PR TITLE
Prepend unique identifier and version in user agent

### PR DESCRIPTION
**Pull request by shrutisinghania**
This PR updates the User-Agent header in requests, allowing for better tracking of the library's usage.

### CODE REVIEW

File: client/src/main/java/com/google/cloud/gcs/analyticscore/client/GcsClientImpl.java [hunk 1]
```diff
@@ -121,6 +121,36 @@ protected Storage createStorage(Optional<Credentials> credentials) {
     return builder.build().getService();
   }
 
+  private static class VersionHolder {
```
Comment (ajayky-os) [inline client/src/main/java/com/google/cloud/gcs/analyticscore/client/GcsClientImpl.java]: We can have a VersionHelper class separately in place of adding a private class within GcsClientImpl, this class seems to have a completely decoupled responsibility.

File: client/src/main/java/com/google/cloud/gcs/analyticscore/client/GcsClientImpl.java [hunk 2]
```diff
@@ -121,6 +121,36 @@ protected Storage createStorage(Optional<Credentials> credentials) {
...
+      return version;
+    }
+  }
+
+  private String getVersion() {
+    return VersionHolder.VERSION;
+  }
+
+  @VisibleForTesting
+  String computeUserAgent() {
```
[the remark this answers is not in the document]
Comment (shrutisinghania) [reply client/src/main/java/com/google/cloud/gcs/analyticscore/client/GcsClientImpl.java]: Renamed the method

File: client/src/test/java/com/google/cloud/gcs/analyticscore/client/GcsClientImplTest.java [hunk 3]
```diff
@@ -209,6 +209,27 @@ void openReadChannel_itemInfoPointsToDirectory_throwsIllegalArgumentException()
         .hasMessageThat()
         .isEqualTo("Expected GCS object to be provided. But got: " + directoryItemId);
   }
+  @Test
+  void getUserAgent_noOptionalUserAgent() {
+    GcsClientImpl client = new GcsClientImpl(TEST_GCS_CLIENT_OPTIONS, executorServiceSupplier);
+    String userAgent = client.getUserAgent();
```
Comment (ajayky-os) [inline client/src/test/java/com/google/cloud/gcs/analyticscore/client/GcsClientImplTest.java]: If spec is simple, that is single line for arrange, act and assert block each then blank line to separate these blocks can be omitted. 

Either remove the blank line between act and assert block or add blank line between arrange and act, keep it consistent.

**Review by ajayky-os — DISMISSED** [review]

**Review by ajayky-os — CHANGES_REQUESTED** [review]

File: client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  @VisibleForTesting
+  static final String DEFAULT_VERSION = "unknown";
+
+  static final String VERSION = loadVersion(PACKAGE_POM_PATH);
+
+  private VersionHelper() {}
+
+  @VisibleForTesting
+  static InputStream getResourceAsStream(String pomPath) {
+    return VersionHelper.class.getResourceAsStream(pomPath);
```
[the remark this answers is not in the document]
Comment (shrutisinghania) [reply client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java]: It is not an infinite loop. `VersionHelper.class.getResourceAsStream(pomPath)` calls the native method available on the `java.lang.Class` instance itself, not this `VersionHelper` method recursively.
The wrapper method exists specifically for testability. Directly mocking `java.lang.Class.getResourceAsStream` is difficult and flaky. By wrapping the call here, the unit test can cleanly use Mockito's `mockStatic(VersionHelper.class)` to mock this method and simulate missing files or `IOException` throws without using complex bytecode manipulation.

File: client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(VersionHelper.class);
+  
+  @VisibleForTesting
+  static final String PACKAGE_POM_PATH =
+      "/META-INF/maven/com.google.cloud.gcs.analytics/client/pom.properties";
+  
+  @VisibleForTesting
+  static final String DEFAULT_VERSION = "unknown";
+
+  static final String VERSION = loadVersion(PACKAGE_POM_PATH);
```
[the remark this answers is not in the document]
Comment (shrutisinghania) [reply client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java]: If `loadVersion()` internally relied on the static constant `PACKAGE_POM_PATH`, the test would be forced to use only that specific resource path. By decoupling the path into a method parameter over passing it inline, the unit test can easily mock paths (like `"dummyPath"`) to verify edge cases.

File: client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    } catch (IOException e) {
+      LOG.warn("Failed to close or access resource stream", e);
+      return DEFAULT_VERSION;
+    }
+  }
+
+  @VisibleForTesting
+  static String loadVersion(InputStream stream) {
+    String version = DEFAULT_VERSION;
+    if (stream != null) {
```
Comment (ajayky-os) [inline client/src/main/java/com/google/cloud/gcs/analyticscore/client/VersionHelper.java]: Use guard clause and return early. 

```
static String loadVersion() {
if (stream == null) { 
  return DEFAULT_VERSION; 
}
try {
  Properties properties = new Properties();
  properties.load(stream);
  return properties.getProperty("version", DEFAULT_VERSION);
} catch (IOException e) {
  LOG.warn(....);
}
return DEFAULT_VERSION;

**Review by ajayky-os — DISMISSED** [review]

**Review by ajayky-os — APPROVED** [review]
